### PR TITLE
Standardize use of `transmit_frequency_start/stop` and use `pulse_form` for BB checks

### PR DESCRIPTION
**Pull request by emiliom**
Addresses #1081 (and has linkage to #951). Specifically:

- For EK80
  - [x] Rename `frequency_start/end` to `transmit_frequency_start/stop`
  - [x] Set attributes to convention specification
  - [x] Change data type from `int` to `float`
  - [x] Overhaul checks for narrowband vs broadband (BB) channels to use `transmit_type` / `pulse_form`.
  - [x] For channels that are not `BB`, set values to `frequency_nominal` rather than `nan`
  - [x] Update checks for narrowband vs BB channels once `transmit_frequency_start/stop` are set to always exist and always have valid, non-`nan` values.
- [x] For EK60, add `transmit_frequency_start/stop` with only a `channel` dimension and set their values to `frequency_nominal`
- [x] For AZFP, add `transmit_frequency_start/stop` with only a `channel` dimension and set their values to `frequency_nominal`

### CODE REVIEW

**Comment by emiliom** [comment]
Based on https://github.com/OSOceanAcoustics/echopype/issues/951#issuecomment-1652198520, the new checks for broadband channels should use `pulse_form > 0` (0 = "CW"). I couldn't find a definition for "FMD" in the EK80 manual; but I assume it's also BB, just like "FM".

**Comment by leewujung** [comment]
I also don't know what "FMD" is, but in the files we have access to it is either 0 or 1.

**Comment by emiliom** [comment]
I'll add the `pulse_form > 0` BB check to this PR once PR #1094 is merged.

**Comment by emiliom** [comment]
Renamed EK80 `pulse_form` to `transmit_type` and changed to string, type per convention.

Not really part of the original scope of this PR, but this change will need to be in place before the new check for BB based on pulse_form (transmit_type) can be implemented.

**Comment by leewujung** [comment]
I went back to the function _get_chan_dict  to check again and think that:
1. like what we discussed, there is a bug if within the channel there was a switch from BB to CW or from CW to BB
2. `np.unique` is pretty expensive, and the way the code was written, it first uses the existence of `frequency_start/end` to determine if there are BB channels (if `frequency_start/end` both exist then there are BB channels in this `Beam_group1`) (more precise, the operation actually checks for if there exists at least 1 BB ping out of all pings across all channels), and then if there is BB channels, drop channels that have all-NaN `freq_center` (this is the case when the BB to CW or CW to BB switch would not be processed correctly). The code is below for reference:
```python
# Use center frequency for each ping to select BB or CW channels
# when all samples are encoded as complex samples
if "frequency_start" in beam and "frequency_end" in beam:
    # At least some channels are BB
    # frequency_start and frequency_end are NaN for CW channels
    freq_center = (beam["frequency_start"] + beam["frequency_end"]) / 2

    return {
        # For BB: drop channels containing CW samples (nan in freq start/end)
        "BB": freq_center.dropna(dim="channel").channel,
        # For CW: drop channels containing BB samples (not nan in freq start/end)
        "CW": freq_center.where(np.isnan(freq_center), drop=True).channel,
    }
```

For 1: I think we should probably document it but not making any changes for the time being. Most of the time things do not get fiddled in the middle of a file, especially because complex data is pretty huge so the number of pings within a typical file, say 100 MB, is pretty small.

For 2.: What do you think about the following:
```python
if 1 in beam["pulse_form"]:
    # At least 1 BB ping exists -- this is analogous to what we had from before
    # Before: when at least 1 BB ping exists, "frequency_start" and "frequency_end" will exist
    freq_center = (beam["frequency_start"] + beam["frequency_end"]) / 2

    # This part below doesn't require changes (so still buggy)
    return {
        # For BB: drop channels containing CW samples (nan in freq start/end)
        "BB": freq_center.dropna(dim="channel").channel,
        # For CW: drop channels containing BB samples (not nan in freq start/end)
        "CW": freq_center.where(np.isnan(freq_center), drop=True).channel,
    }

```

**Comment by emiliom** [comment]
> 1. like what we discussed, there is a bug if within the channel there was a switch from BB to CW or from CW to BB

Quick note: by coincidence, I discovered that this raw file may be one such case: `TEST_DATA_FOLDER / "ek80/2019118 group2survey-D20191214-T081342.raw"`

**Comment by emiliom** [comment]
> For 2.: What do you think about the following:
```python
if 1 in beam["pulse_form"]:
   # At least 1 BB ping exists -- this is analogous to what we had from before
   # Before: when at least 1 BB ping exists, "frequency_start" and "frequency_end" will exist
```

That'll work. Come to think of it, my use of `np.unique` was unnecessary, and not just in this particular check! 

For this check, though, I know you've said that you haven't encountered the "FMD" (`pulse_form` = 5) type, but are you saying we should omit checking for it? I realize adding the `5 in beam["pulse_form"]` to the `if` condition adds to the computation, but it seems to me that if it's in the Simrad EK80 manual, we should be comprehensive and account for it. No?

Now, back to `np.unique`. There are 5 places, in 4 modules, where a test for BB vs CW occurs, and these checks come in two types: 

**1. The one above from `calibrate_ek.py` (below) and two occurrences, which test for the occurrence of a BB ping:**

https://github.com/OSOceanAcoustics/echopype/blob/eb92185e83577b3e30bdcaf00e3e43ec0ada547a/echopype/calibrate/calibrate_ek.py#L314-L316

https://github.com/OSOceanAcoustics/echopype/blob/eb92185e83577b3e30bdcaf00e3e43ec0ada547a/echopype/echodata/simrad.py#L141-L145

https://github.com/OSOceanAcoustics/echopype/blob/eb92185e83577b3e30bdcaf00e3e43ec0ada547a/echopype/convert/set_groups_ek80.py#L647-L653

**2. A test for the *absence* of BB pings (ie, all CW), with two occurrences:**

https://github.com/OSOceanAcoustics/echopype/blob/eb92185e83577b3e30bdcaf00e3e43ec0ada547a/echopype/echodata/simrad.py#L121-L126

https://github.com/OSOceanAcoustics/echopype/blob/eb92185e83577b3e30bdcaf00e3e43ec0ada547a/echopype/calibrate/ek80_complex.py#L214-L218

I can now see that this test can be handled with the `not in` operator: `1 not in beam[`pulse_form`] and 5 not in beam[`pulse_form`]`.

Ok, I think the path is clear now. Thanks!

**Comment by leewujung** [comment]
> > For 2.: What do you think about the following:
> 
> ```python
> if 1 in beam["pulse_form"]:
>    # At least 1 BB ping exists -- this is analogous to what we had from before
>    # Before: when at least 1 BB ping exists, "frequency_start" and "frequency_end" will exist
> ```
> 
> That'll work. Come to think of it, my use of `np.unique` was unnecessary, and not just in this particular check!
> 
> For this check, though, I know you've said that you haven't encountered the "FMD" (`pulse_form` = 5) type, but are you saying we should omit checking for it? I realize adding the `5 in beam["pulse_form"]` to the `if` condition adds to the computation, but it seems to me that if it's in the Simrad EK80 manual, we should be comprehensive and account for it. No?

Good point. I guess then it'll be the reverse of the previous code, making it like below:
```
if not np.all(beam["pulse_form"] == 0):
```

**Comment by leewujung** [comment]
> > 1. like what we discussed, there is a bug if within the channel there was a switch from BB to CW or from CW to BB
> 
> Quick note: by coincidence, I discovered that this raw file may be one such case: `TEST_DATA_FOLDER / "ek80/2019118 group2survey-D20191214-T081342.raw"`

Oh! ok I'll take a look. Seems that we've not used this file for cal tests, but I suspect that the operations will go through, but the results will be wrong. I can create a new issue for this.

**Comment by emiliom** [comment]
> Good point. I guess then it'll be the reverse of the previous code, making it like below:
> if not np.all(beam["pulse_form"] == 0):

Nicer and cleaner. But wouldn't it be more expensive computationally than this?
```python
1 in beam["pulse_form"] or 5 in beam["pulse_form"]
# or, more compactly:
np.isin(beam["pulse_form"], [1,5]).any()
```

**Comment by emiliom** [comment]
>> Quick note: by coincidence, I discovered that this raw file may be one such case: TEST_DATA_FOLDER / "ek80/2019118 group2survey-D20191214-T081342.raw"
>
> Oh! ok I'll take a look. Seems that we've not used this file for cal tests, but I suspect that the operations will go through, but the results will be wrong. I can create a new issue for this.

That's correct: the operations went through. I was using that file (don't remember why I picked it) in my echopype checker, when I noticed the results.

**Comment by emiliom** [comment]
I've implemented the checks for BB based on `transmit_type` / `pulse_form`. 

The last remaining task is: For channels that are not `BB`, set values to `frequency_nominal` rather than `nan`. Then, the checks for BB may need to receive small updates.

Note that I haven't yet merged into this PR the changes from PR's #1099 and #1096.

**Comment by emiliom** [comment]
Ok, this PR is now complete and ready for review (once the CI completes sucessfully)!

**Comment by emiliom** [comment]
A note: the comments above about ek80 `pulse_form` all use the integer codes (0,1,5). Now, with the changeover to the convention-based `transmit_type`, these are strings: "CW", "FM", "FMD"

File: echopype/calibrate/ek80_complex.py
Comment (leewujung) [inline echopype/calibrate/ek80_complex.py:15]: seems like `transmit_type` is unused and can be removed

File: echopype/calibrate/ek80_complex.py
Comment (leewujung) [inline echopype/calibrate/ek80_complex.py:239]: In theory this can be removed since it is not used in `tapered_chirp`, but I kinda like having the uniqueness check on `transmit_type`. How about adding the below comment?
```suggestion
        "transmit_type",  # not needed for tapered_chirp but good to check uniqueness
```

File: echopype/convert/set_groups_ek80.py
Comment (leewujung) [inline echopype/convert/set_groups_ek80.py:649]: I like your change. Maybe we can remove the commented out line?

File: echopype/convert/set_groups_ek80.py
Comment (leewujung) [inline echopype/convert/set_groups_ek80.py:938]: Should we change this to "Frequency Modulated user-Defined" or do you think this is extrapolating too far from the documentation about ADCP?

File: echopype/echodata/simrad.py
Comment (leewujung) [inline echopype/echodata/simrad.py:127]:
```suggestion
        # when echodata contains CW power and BB complex samples.
```
I think we can safely remove the rest of the sentence since now `transmit_frequency_start` would always exist.

File: echopype/echodata/simrad.py
Comment (leewujung) [inline echopype/echodata/simrad.py:148]: I think there is a logic error in the first if here. `transmit_type_has_BB` says whether or not there is one ping that is not CW, but does not tell us whether there is no CW ping at all (which is the case we want to catch, and is in the second if. I think this can be:
```python
        # Raise error if waveform_mode="CW" but CW data does not exist (not a single ping is CW)
        if encode_mode == "complex" and np.all(echodata["Sonar/Beam_group1"]["transmit_type"] != "CW"):
            raise ValueError("waveform_mode='CW', but all data are broadband (BB)!")
```

File: echopype/echodata/simrad.py
Comment (leewujung) [inline echopype/echodata/simrad.py:123]: This is probably an expensive operation across all `ping_time`. Since right now we have this bug that we can only correctly process all-CW channels or all-BB channels (but not CW-BB mixed _within_ the same channel), do you think it's a bad idea to just test for `transmit_type` of the first ping?

File: echopype/calibrate/calibrate_ek.py
Comment (leewujung) [inline echopype/calibrate/calibrate_ek.py:323]: We do not need `freq_center` here anymore now that `transmit_type` is available.

The `BB: ...` and `CW: ...` selections do not work if there are within-channel changes of CW and FM. In those cases, either only the single elements fits the criteria are kept, or only the single elements fits the criteria are set to nan. For example, if you have such within-channel changes:
Screenshot 2023-08-02 at 5 36 33 PM
Then the following happens:
first selection:
Screenshot 2023-08-02 at 5 37 54 PM
second selection:
Screenshot 2023-08-02 at 5 36 48 PM

I think we can do the following
```python
first_ping_transmit_type = beam["transmit_type"].isel(ping_time=0)  # assume transmit_type identical for all pings
return {
    # For BB: Keep only non-CW channels (FM or FMD) based on transmit_type
    "BB": first_ping_transmit_type.where(first_ping_transmit_type != "CW", drop=True).channel,
    # For CW: Keep only CW channels based on transmit_type
    "CW": first_ping_transmit_type.where(first_ping_transmit_type == "CW", drop=True).channel,
}
```

**Review by leewujung — CHANGES_REQUESTED** [review]
@emiliom : Thanks for the PR! I think there are a couple places that have bugs. Could you look through my comments and change them? 

I think we will have to add a disclaimer in the documentation that the code does not currently handle within-channel FM-CW switches if they are both encoded in complex samples. (I still need to write an issue to keep track of this.)

File: echopype/calibrate/ek80_complex.py
Comment (emiliom) [reply echopype/calibrate/ek80_complex.py:15]: Right, thanks! I'll include this in the next batch of commits.

File: echopype/calibrate/ek80_complex.py
Comment (emiliom) [reply echopype/calibrate/ek80_complex.py:239]: Right! I added it for an initial implementation that used `transmit_type` in `tapered_chirp`, then removed that usage (simplified the code) and forgot to remove the unused `transmit_type`.

Sure, I'll add that comment in the next set of commits.

File: echopype/convert/set_groups_ek80.py
Comment (emiliom) [reply echopype/convert/set_groups_ek80.py:649]: Sigh, thanks. Will do.

File: echopype/echodata/simrad.py
Comment (emiliom) [reply echopype/echodata/simrad.py:127]: Right, thanks. Will do.

File: echopype/convert/set_groups_ek80.py
Comment (emiliom) [reply echopype/convert/set_groups_ek80.py:937]: I don't know. As I mentioned in https://github.com/OSOceanAcoustics/echopype/pull/1101#issuecomment-1657451416, even the description next to LFMD seems scrambled and unclear. Having the acronyms spelled out here, as far as we can, seemed like progress, but I'm hesitant to go further. Still, these are only attribute values that would have no downstream implications, so they can be easily tweaked in the future.

File: echopype/echodata/simrad.py
Comment (emiliom) [reply echopype/echodata/simrad.py:147]: I think it's not so much that the logic is in error (or led to errors), but rather that having `transmit_type_has_BB` in the first if is unnecessary and doesn't add anything useful. I now see that I translated the original logic too literally, but didn't take the further step you did. I agree with your suggestion and will implement it. Thanks! FYI, I'll also remove the variable `transmit_type_has_BB` since it will no longer serve the purpose of reusability.

File: echopype/echodata/simrad.py
Comment (emiliom) [reply echopype/echodata/simrad.py:123]: Surprisingly (at least to me), it's no more expensive than the `"CW" in dataarray` scheme we adopted elsewhere. See the tests I reported above, https://github.com/OSOceanAcoustics/echopype/pull/1091#issuecomment-1656841321

> Since right now we have this bug that we can only correctly process all-CW channels or all-BB channels

That's true, but the test is done across *all* channels, not for an individual channel. So, while the first ping tells us about the channel it corresponds to, it doesn't tell us anything about other channels.

Comment (emiliom) [reply echopype/echodata/simrad.py:123]: On second thought, maybe I wasn't seeing the multi-dimensional aspect of the data .... The "first" ping is the first ping in *each* channel. In that case, I guess you're right.

File: echopype/calibrate/calibrate_ek.py
Comment (emiliom) [reply echopype/calibrate/calibrate_ek.py:323]: I follow the logic and it seems like it'd work. But I ran into multiple errors when testing it. I'll have to go into debug mode on it tomorrow morning, to look at it more carefully.

File: echopype/calibrate/ek80_complex.py
Comment (emiliom) [reply echopype/calibrate/ek80_complex.py]: I take this back. `transmit_type` can't remain in `tx_params` b/c the `tx_params` dict is passed to `tapered_chirp`, but `transmit_type` is no longer an argument in that function.

File: echopype/calibrate/calibrate_ek.py
Comment (emiliom) [reply echopype/calibrate/calibrate_ek.py]: Alright, I found and fixed the problem. The first statement in your suggestion retained the `ping_time` variable, which needs to be removed. So, the fixed statement is:
```python
first_ping_transmit_type = beam["transmit_type"].isel(ping_time=0).drop_vars("ping_time")
```

File: echopype/echodata/simrad.py
Comment (emiliom) [reply echopype/echodata/simrad.py]: I've implemented a solution based on the first ping_time.

File: echopype/echodata/simrad.py
Comment (emiliom) [reply echopype/echodata/simrad.py]: This change is in the latest commits.

File: echopype/convert/set_groups_ek80.py
Comment (emiliom) [reply echopype/convert/set_groups_ek80.py:956]: Let's leave it as is until we have better information.

**Comment by emiliom** [comment]
I've addressed all your comments. Thanks so much! Hopefully the CI will complete w/o issues.

I'm not sure if the proper Github step here is to "Dismiss" your previous review, or if you should do that after reviewing my changes. I requested another review, but maybe that was unnecessary. Oh well.

**Comment by emiliom** [comment]
Woo-hoo, the CI tests all passed!

File: echopype/echodata/simrad.py
Comment (leewujung) [inline echopype/echodata/simrad.py:144]: Since we only test for `first_ping_transmit_type` in the BB case, to make things more clear when we pick up the issue of within-channel CW-BB switches, maybe we change the check below to just the first ping too?
```suggestion
        # Raise error if waveform_mode="CW" but CW data does not exist (not a single ping is CW)
        # TODO: this needs to be changed when fixing within-channel CW-BB switch
        if encode_mode == "complex" and np.all(first_ping_transmit_type != "CW"):
            raise ValueError("waveform_mode='CW', but all data are broadband (BB)!")
```

File: echopype/echodata/simrad.py
Comment (leewujung) [inline echopype/echodata/simrad.py:124]: Just adding a TOOD:
```suggestion
    # assume transmit_type identical for all pings in a channel
    # TODO: change when allowing within-channel CW-BB switch
```

**Review by leewujung — APPROVED** [review]
LGTM, thanks! I suggested a small tweak to just to make things parallel when we get around to fix the within-channel CW-BB switch.

File: echopype/echodata/simrad.py
Comment (emiliom) [reply echopype/echodata/simrad.py:144]: Looks good. I'll accept the suggestion here rather than via a push from me.

**Comment by emiliom** [comment]
Thanks! I committed your two inline suggestions here. As soon as the CI is done, I can merge.

**Comment by emiliom** [comment]
I've pushed the changes we discussed:
- FM -> LFM
- More fleshed out `flag_meanings`, largely from the convention descriptions

Assuming everything checks out, we can then merge this PR

**Comment by leewujung** [comment]
Looks good. I'll merge now!

**Pull request merged**
